db/compaction/compaction_picker_level.cc

```diff
@@ -773,7 +773,17 @@ bool LevelCompactionBuilder::PickFileToCompact() {
 
   const std::vector<FileMetaData*>& level_files =
       vstorage_->LevelFiles(start_level_);
-
+  // if total size of Level-0 is small than target_file_size_base, do not pick file to 
```

Comment:
In the case when no intra-L0 compaction is picked, I worry we will be in a state with too many L0 files. Can we continue to pick an L0 file if we know that no intra-L0 compaction will be picked?

Reply:
> In the case when no intra-L0 compaction is picked, I worry we will be in a state with too many L0 files. Can we continue to pick an L0 file if we know that no intra-L0 compaction will be picked?

if `PickFileToCompact` return false, when start_level_ is 0, [PickIntraL0Compaction will be picked](https://github.com/facebook/rocksdb/blob/main/db/compaction/compaction_picker_level.cc#L213) to compaction

Reply:
I mean that when `PickFileToCompact()` return false here due to L0 being too small, and then if `PickIntraL0Compaction()` also returns false, we would end up with more L0 files compared to the existing behavior.

Reply:
I also change the  `kMinFilesForIntraL0Compaction` to 2 to make sure  that`PickIntraL0Compaction `always return true only when  the size of L0 files is greater than `(level0_file_num_compaction_trigger + 2)`  or some L0 file is compacting. if L0 file is not compacting, the size of L0 file will never over  `(level0_file_num_compaction_trigger + 2) `

Reply:
Changing `kMinFilesForIntraL0Compaction` may be risky as it allows more L0 compaction when l0 file is not small. How about we just call `PickIntraL0Compaction()` here if `total_l0_size < mutable_cf_options_.target_file_size_base` and pass in a flag to ignore the check `level_files.size() <
static_cast<size_t>(mutable_cf_options_.level0_file_num_compaction_trigger + 2)` (maybe still ensure that we have at least `level0_file_num_compaction_trigger` L0 files). Return true if an intra-l0 compaction is found.